File: src/main/java/frc/team449/robot2022/drive/DriveConstants.java
package frc.team449.robot2022.drive;

import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.util.Units;
import edu.wpi.first.wpilibj.Encoder;

public class DriveConstants {

  /** Drive motor ports */
  public static final int DRIVE_MOTOR_FL = 0;
  public static final int DRIVE_MOTOR_FR = 1;
  public static final int DRIVE_MOTOR_BL = 2;
  public static final int DRIVE_MOTOR_BR = 3;
  public static final int TURN_MOTOR_FL = 4;
  public static final int TURN_MOTOR_FR = 5;
  public static final int TURN_MOTOR_BL = 6;
  public static final int TURN_MOTOR_BR = 7;

  /** External encoders for driving motors */
  public static final Encoder DRIVE_ENC_FL = new Encoder(0, 1);
  public static final Encoder DRIVE_ENC_FR = new Encoder(2, 3);
  public static final Encoder DRIVE_ENC_BL = new Encoder(4, 5);
  public static final Encoder DRIVE_ENC_BR = new Encoder(6, 7);

  /** Analog encoder channels */
  public static final int TURN_ENC_CHAN_FL = 0;
  public static final int TURN_ENC_CHAN_FR = 1;
  public static final int TURN_ENC_CHAN_BL = 2;
  public static final int TURN_ENC_CHAN_BR = 3;

  /** Offsets for the absolute encoders */
  public static final double TURN_ENC_OFFSET_FL = 0;
  public static final double TURN_ENC_OFFSET_FR = 0;
  public static final double TURN_ENC_OFFSET_BL = 0;
  public static final double TURN_ENC_OFFSET_BR = 0;

  public static final double DRIVE_WHEEL_RADIUS = Units.inchesToMeters(2);
  public static final double DRIVE_GEARING = 5.86;
  // todo determine this
  public static final double TURN_GEARING = 1;
  public static final double DRIVE_UPR = 0.3021211527151539;
  // todo determine this. should be in radians
  public static final double TURN_UPR = 2 * Math.PI * 1;
  public static final int NEO_ENCODER_CPR = 1;
  /** CPR of external encoders on driving motors */
  public static final int DRIVE_EXT_ENC_CPR = 256;

  /** Location of the front left module */
  public static final Translation2d FRONT_LEFT_LOC = new Translation2d(
    Units.inchesToMeters(25) / 2,
    Units.inchesToMeters(25) / 2
  );

  public static final int DRIVE_CURRENT_LIM = 40;
  public static final double DRIVE_ENC_VEL_THRESHOLD = 0.1;
  public static final double MAX_LINEAR_SPEED = 4, MAX_ROT_SPEED = 1;
}

File: src/main/kotlin/frc/team449/robot2022/drive/DriveConstants.kt
package frc.team449.robot2022.drive

import edu.wpi.first.apriltag.AprilTag
import edu.wpi.first.apriltag.AprilTagFieldLayout
import edu.wpi.first.math.geometry.Pose2d
import edu.wpi.first.math.geometry.Pose3d
import edu.wpi.first.math.geometry.Rotation2d
import edu.wpi.first.math.geometry.Transform3d
import edu.wpi.first.math.util.Units

object DriveConstants {

  /** Drive motor ports */
  const val DRIVE_MOTOR_FL = 1
  const val DRIVE_MOTOR_FR = 2
  const val DRIVE_MOTOR_BL = 3
  const val DRIVE_MOTOR_BR = 4

  /** Feed forward values for driving each module */
  const val DRIVE_KS = 0.16475
  const val DRIVE_KV = 2.0909
  const val DRIVE_KA = 0.29862

  /** PID gains for driving each module*/
  const val DRIVE_KP = 0.35
  const val DRIVE_KI = 0.0
  const val DRIVE_KD = 0.0

  /** Drive configuration */
  const val DRIVE_GEARING = 1 / 8.0
  val DRIVE_UPR = Math.PI * Units.inchesToMeters(6.0)
  const val MAX_LINEAR_SPEED = 1.5 // 1.5
  const val MAX_ROT_SPEED = .1 // m'/s
  const val MAX_ATTAINABLE_MK4I_SPEED = (12 - DRIVE_KS) / DRIVE_KV
  const val MAX_ACCEL = 5.0

  /** Controller Configurations */
  const val RATE_LIMIT = 1.5
  const val TRANSLATION_DEADBAND = .125
  const val ROTATION_DEADBAND = .125

  val CAM_NAME = "limelight"
  val ROBOT_TO_CAM = Transform3d()
  val TAG_LAYOUT = AprilTagFieldLayout(listOf(AprilTag(0, Pose3d())), 16.4846, 8.1026)

  val INITAL_POSE = Pose2d(0.0, 0.0, Rotation2d.fromDegrees(-180.0))

  val WHEELBASE = Units.inchesToMeters(21.426)
  val TRACKWIDTH = Units.inchesToMeters(21.000)
}
